var/spack/repos/builtin/packages/boost/package.py

```diff
@@ -624,9 +624,14 @@ def determine_b2_options(self, spec, options):
                 options.append("runtime-link=shared")
             else:
                 options.append("runtime-link=static")
-            for lib in self.all_libs:
-                if f"+{lib}" not in spec:
-                    options.append(f"--without-{lib}")
+
+            # Any lib that is in self.all_libs and in the variants dictionary
+            # and is set to False should be added to options in a --without flag
+            all_lib_set = set(self.all_libs) & set(self.spec.variants.dict.keys())
+            all_libs_off = filter(
+                lambda v: v.name in all_lib_set and not v.value, self.spec.variants.dict.values()
+            )
+            options.extend([f"--without-{lib.name}" for lib in all_libs_off])
```

Comment:
I thought originally we spoke about removing the `charconv` variant entirely on windows.

I'm not clear on how this solution works, for example

```
filter(
                lambda v: v.name in all_lib_set and not v.value, self.spec.variants.dict.values()
            )
```

seems the same as

```
[x for x in all_lib_set if f"+{x}" not in spec]
```

and as for 

```
all_lib_set = set(self.all_libs) & set(self.spec.variants.dict.keys())
```

I'm not clear what that's doing because `all_libs` is used to instantiate variants, i.e. it should be that `set(all_lib_set) == set(self.all_libs)`

Reply:
@scheibelp - 

> I thought originally we spoke about removing the charconv variant entirely on windows.

`charconv` was added to boost@1.85.0 and it builds fine on windows with that version. More info below.

> I'm not clear on how this solution works

**The problem** 

- `charconv` was added to `all_libs` list for every version; it is only available starting in boost@1.85.0. 
- The old way of constructing the `--without-<lib>` list added *any lib*  in `all_libs` so long as `not spec.satisfies("+<lib>")`. 
- This resulted in a `b2` config error in boost@1.82.0 due to unrecognized `--without-charconv` flag.

**The solution** 

- only add `--without` flags of libs that are in `all_libs` AND allowed variants. 
- To construct this list, we need to find all variants (libs) that match the following pseudocode logic:
1. the lib/variant must be a valid `--without-<lib>` flag
```
<lib> in self.all_libs == True
```
2. the lib/variant is allowed by the specific version/os/compiler/etc.
```
<lib> in self.spec.variants.dict == True
```
3. the lib/variant is turned off
```
self.spec.variants.dict[<lib>].value == False
```

This is working locally for me for newer (1.85.0) and older (1.82.0) versions of boost.